### ISSUE TITLE
franken-moe: --dfa-feedback kolen-pollack — adaptive/learnable expert feedback that tracks the forward weights (Tao F13, the untested lever)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#150]
The single biggest untested variable in the whole DFA-vs-BP program: **every DFA experiment (F4–F9m) used a FIXED random feedback matrix B** (`--dfa-b-seed`), never updated. The recent literature review + ICLR2026 work makes clear that the DFA methods which actually work at LLM-adjacent scale do NOT use fixed random feedback — they make the feedback **adaptive** (learnable, tracking the forward weights), which is exactly the ingredient our own `align` metric says we're missing.

## Why (mechanistically, and it's our own data)

Our `align` events already measure `cos(g_dfa, g_bp)` per step. The adversarial review found this near-ZERO at vocab 50257 (it scales inversely with output dim). The literature's explanation for DFA working at all: *forward weights rotate toward the feedback subspace → alignment rises → useful updates*. **Direct Kolen–Pollack (DKP)** and related adaptive-feedback methods ACTIVELY drive that — they update B each step to track the forward path, and beat fixed-DFA (CIFAR100/TinyImageNet). ICLR2026 "Feedback-Weight Matching" makes the same point for fine-tuning: DFA needs the `W ∝ F Fᵀ` alignment condition, which fixed random B never satisfies.

So fixed-B DFA (what we tested, and found counterproductive) and adaptive-B DFA (what the literature rides) are **different recipes**. We've never tested the second.

## Ask

A feedback mode selector on the DFA-experts path:

```
--dfa-feedback fixed|kolen-pollack     # default fixed (must be byte-null vs absent)
--dfa-feedback-decay LAMBDA            # weight-decay coefficient for the KP coupling (e.g. 1e-3)
--dfa-feedback-lr ETA                  # optional: separate LR for the B update (default: tie to --lr)
```

Under `kolen-pollack`, the expert feedback matrix B is UPDATED each step (instead of frozen) by a Kolen–Pollack coupling: B receives the transposed forward update plus weight decay, so B converges toward the (effective) forward path and `cos(g_dfa, g_bp)` rises. The precise rule is your call — the canonical references are **Akrout et al. 2019 "Deep Learning without Weight Transport"** (KP + weight decay drives feedback→forward alignment) and the **Direct Kolen–Pollack** DFA variant. For our direct (output→expert) feedback, the design choice is whether B tracks the immediate forward weight or the effective output-path product; pick what's cleanest and note it.

## Telemetry (this is the whole point)
- The existing `align` event (`cos`, `dfa_norm`, `bp_norm`) must work in this mode — under KP we expect `cos` to RISE over training, vs staying ~0 under `fixed`. That trend IS the result.
- Add a feedback-tracking metric to align or run_end: `cos(B, forward)` (how aligned the feedback has become), so we can separate "B moved" from "B moved toward the forward weights."
- Provenance: `dfa_feedback` mode + decay/lr in `run_start.franken_moe`.

## Gate expectations
- `--dfa-feedback fixed` byte-null vs absent (structural, like the lr-control none).
- `kolen-pollack`: B provably MOVES (not frozen); `cos(B, forward)` provably RISES over a short run (the coupling works); `align` `cos` higher than the fixed twin by end; deterministic (seed-equal byte-identical); guards (decay ≥ 0, kolen-pollack requires dfa-experts).
- Composes with `--dfa-granularity block`, `--shape deep`, `--routing dense`.
- No claim about final loss — that's F13's science, not the gate's.

## Not asking for
FWM-style aligned RE-INIT of a pretrained spine (the ICLR2026 fine-tuning recipe) — that's a separate surface for the F12 retrofit story; will file separately if F13 shows adaptive feedback has legs.

Consuming experiment: **Tao F13** — re-run the F9d control (frozen vs fixed-DFA vs BP at the contested fixture) with a fourth `dfa-kp` arm. If adaptive feedback raises alignment AND beats frozen where fixed-B lost, the program's negative was fixed-B-specific and DFA-as-subordinate revives (correctly). If alignment rises but loss still loses to frozen, that's a sharper negative. Either way it's the first new mechanism since the block-DFA line.

**Comment by OriPekelman** [comment]
Started. Design is settled and committed (`30e481c`, battery green 54/54); the mechanism itself is **not** built yet. Full writeup: `docs/roadmap/toy150-kolen-pollack-design-2026-08-06.md`.

## The rule I picked (you delegated this — "pick what's cleanest and note it")

`B_l` stands in for the **effective output path** `P_l`: the map from the activation `B_l` feeds back into, forward to the logits. The loss gradient w.r.t. that path is exactly

```
∇_P L = e · a_outᵀ
```

so the coupling is `B ← B − η(e·a_outᵀ) − ηλB` — Kolen–Pollack with weight decay. The decay is what makes feedback and path *converge* rather than merely co-move.

**This turns out to be nearly free:** that is precisely ggml's SGD step, `opt_step_sgd(B, g, [η, λ])`. No new kernel, no new optimizer arm — `B` simply stops being an inert persistent input.

**On the choice you flagged** ("immediate forward weight vs effective output-path product"): the immediate-weight option doesn't typecheck. The feedback is *direct*, so `B_l` is `[vocab, d_out]` while the immediate forward weight is `[d_in, d_out]` — no shape makes one mirror the other. The path formulation is the only one whose dimensions close, and it's what the DKP literature uses for direct feedback. So the choice was made for us.

## What the shape audit turned up

| B | ne | `a_out` needed |
|---|---|---|
| `b_ups` | `[vocab, dfv]` | `tap_as` ✅ |
| `b_glus` | `[vocab, dfv]` | `tap_as` ✅ |
| `b_rs` | `[vocab, nev]` | `t_gates` ✅ |
| `b_downs` | `[vocab, ew_b]` | `o_i` ❌ **not tapped** |

That gap is what `30e481c` closes — `tap_os`, parallel to `tap_as`. Inert and byte-null; nothing reads it yet.

## One correction on the telemetry ask

You asked for `cos(B, forward)` to separate "B moved" from "B moved toward the forward weights". Worth noting: **the existing `align` event already does that.** `cos(g_dfa, g_bp)` is high exactly when `B ≈ P`, because `g_dfa = a_inᵀ(Be)ᵀ`. Its rising over training *is* the result you're after — no new metric required for that separation.

I'll still add two things on top:

- **`dfa_b_sig`** — sum of squares over every `B`, in `run_start`/`run_end`. Proves `B` moves at all, and must be **bit-identical** under `fixed`. That's the K4b/`experts_sig` discipline: a coupling that silently never fires looks identical on a loss curve.
- **`b_cos_head`** — a true `cos(B, P)`, but honestly computable in exactly **one** place: the *last* layer's `b_downs`, whose path to the logits is just the tied head. Every earlier layer has intervening blocks in its path, so I'll label it as scoped to that matrix rather than present it as global alignment. (Also only valid when `ew_b == dm`, i.e. not under `--moe-latent`.)

## Remaining

Sequenced in the doc: env + guards → the persistent `t_hp_fb` → update nodes → CLI flags + matrix row → provenance (recording *which* rule ran) → gate leg. Gate will pin `fixed` byte-null, `dfa_b_sig` bit-identical under `fixed` and moved under `kolen-pollack`, `b_cos_head` rising, determinism, composition with block/deep/dense, and the guards — **no claim about final loss**, that's F13's to make.

One landmine already noted for whoever picks it up: `t_hp_fb` must be allocated **persistent, before `finalize_weights`**. A compute-context input allocated later reads zeros in silence (toy#133), which here would mean η=0 — a KP arm that quietly never adapts and looks like a clean negative.

Not blocked on anything; just needs another session.

**Comment by OriPekelman** [comment]
Landed in `6e5616b`. `make gates` 54/54 green. **Read the two caveats before launching F13** — one changes your gains, one is a gap against what you asked for.

## The rule (as designed, now built)

`B ← B − η(e·a_outᵀ) − ηλB`, where `∇_P L = e·a_outᵀ` is the gradient of the effective output path `B` stands in for. Implemented as ggml's SGD step, so no new kernel. Provenance records `dfa_feedback_rule = "path"` so a bundle says which rule ran.

```
--dfa-feedback fixed|kolen-pollack      # default fixed, byte-null
--dfa-feedback-decay LAMBDA             # default 1e-3
--dfa-feedback-lr ETA                   # default: tied to --lr  <-- SEE BELOW
```

## ⚠️ 1. The default gain diverges — do not use it

With η tied to `--lr` (0.02), ‖B‖² grows **1000× in 20 steps** and `cos(B, head)` does *not* rise. That's divergence, not tracking. At sane gains it does exactly what your hypothesis predicts:

| η | λ | ‖B‖² ratio | cos(B, head) |
|---|---|---|---|
| **0.02 (the default)** | 1e-3 | **1000×** | 0.000000 → −0.000711 |
| 1e-4 | 0.01 | 1.00× | 0.000000 → **+0.001675** |
| 1e-6 | 0.01 | 1.00× | 0.000000 → **+0.002132** |
| 1e-8 | 0.05 | 1.00× | 0.000000 → **+0.001687** |

I kept tie-to-`--lr` as the default because you specified it, rather than silently substituting my own — but **F13 should start at `--dfa-feedback-lr 1e-4 --dfa-feedback-decay 0.01`**. `dfa_b_sig` is in every bundle's `run_start`/`run_end`, so if a cell diverges you'll see it immediately instead of inferring it from a bad curve.

(Also fixed on the way: the CLI rejected scientific notation, so `--dfa-feedback-lr 1e-4` would have been refused outright.)

## ⚠️ 2. `--dfa-granularity block` is NOT covered — it fails loud

You asked for composition with block-DFA. It isn't there, and I want to be explicit about why rather than let you find out from a result.

I wired the coupling into the block lane. It built, ran cleanly, 20 steps, no error — and produced a curve **byte-identical to fixed-B**. The block feedback matrix was never actually stepped. Relocating the update after `add_to_graph` didn't help either. The block lane builds its loss root differently (`b_blks` + `full_sum` terms) and the step nodes don't attach the way they do on the matmul path.

An adaptive arm that silently doesn't adapt is the worst possible outcome for F13 — you'd get a confident negative that means nothing. So the combination now **rejects**:

```
--dfa-feedback kolen-pollack does not yet reach --dfa-granularity block
(the block feedback matrix is not coupled; the arm would run and silently NOT adapt).
Use the default matmul granularity.
```

The gate pins that rejection, so it can't regress into a silent no-op. **This ticket stays open for it.** If F13's design needs block-DFA specifically, say so and I'll take it next — it needs a proper look at how `b_blks`' root attaches, not a guess.

For now F13 can run the adaptive arm at **matmul granularity**, which is the granularity F4–F9d used anyway.

## Telemetry

- **`dfa_b_sig`** — sum of squares over the feedback matrices, `run_start` + `run_end`. Bit-identical under `fixed`, moves under `kolen-pollack`. This is the load-bearing check: a coupling that never fires produces a perfectly healthy loss curve.
- **`dfa_b_cos_head`** — a true `cos(B, P)`, scoped to the one place it's honest: the last layer's down feedback, whose path to the logits *is* the tied head. Labelled as scoped, not sold as global alignment.
- **Your existing `align` event already answers the main question** — `cos(g_dfa, g_bp)` is high exactly when `B ≈ P`, so its rising over training *is* the result. No new metric needed for that separation.

## Gate

`fixed` byte-null **and** `dfa_b_sig` bit-identical (the frozen control is verified, not assumed); B provably moves *and stays bounded*; `cos(B, head)` rises; matched init across arms; provenance names rule and gains; the block-DFA rejection; determinism; 4 guards. **No assertion about final loss** — that's yours to make.

**Comment by OriPekelman** [comment]
**Block-DFA composition now works** — `29e4639`, battery green 54/54. The gap I flagged in the previous comment is closed; **F13 can run the adaptive arm at block granularity.**

## The bug was placement, not the mechanism

`tnn_extend_backward_graph` has nothing to attach to before `tnn_build_backward`. I had wired the block coupling inside the surrogate-root loop (~L1247), but `build_backward` is at L1274 and the expert-lane wiring at L1402 — *after* it. The step nodes were silently discarded, which is why it built, ran 20 clean steps, and produced a curve byte-identical to fixed-B. Moving it to just after `build_backward` fixes it:

```
block-DFA:  base 1c9d8d872561  ->  kp acf45848b4e7   (20 steps)
```

Nothing about that failure was visible at the call site, so the placement now carries a comment explaining it.

## A second defect the fix exposed — worth knowing about

With the coupling working, `dfa_b_sig` **still** reported `5120.29 → 5120.29`, i.e. "B frozen", on an arm whose curve provably moved.

Under block-DFA the coupled matrix is `b_blks`; the expert B's aren't used at all (the experts are chain-wired from the surrogate root). `feedback_sig` summed only the expert B's. So the bundle would have **lied** about whether the coupling fired — exactly what `dfa_b_sig` exists to prevent. `dfa_b_cos_head` had the same problem, reading `b_downs`.

Both now follow the matrix that is actually coupled:

```
block+KP:  b_sig 6656.10 -> 6690.40  (1.005x, bounded)
           |cos(B, head)| 7.0e-5 -> 1.23e-4  (rises)
```

Note the implication for reading results: **`dfa_b_sig` values are not comparable between the two granularities** (block includes `b_blks`, matmul covers the expert B's), so compare within an arm's own start→end, not across granularities.

## Gate

Now asserts real composition instead of the rejection: curve differs under block, `dfa_b_sig` moves **and stays bounded**, `cos(B, head)` rises. `fixed` is still byte-identical to the pre-toy#150 baseline.

## Still standing from the previous comment

**The default gain diverges.** η tied to `--lr` (0.02) grows ‖B‖² 1000× in 20 steps. Start F13 at `--dfa-feedback-lr 1e-4 --dfa-feedback-decay 0.01`. That's unchanged and still the thing most likely to cost you a run.

## One flake, recorded rather than waved through

An intermediate battery reported `block-dfa: explicit --dfa-granularity matmul differs from the default (flag-null broken)`. It did **not** reproduce: byte-identical on repeated direct runs at the gate's exact wide/E8 config, the `franken-moe` gate passes standalone, and the full battery has been green twice since. I've left it unexplained rather than claiming it fixed. If you ever see a byte-equality leg fail once and pass on retry, tell me — that would make it real nondeterminism in the matmul lane, and it would take priority over anything else here.

**Comment by OriPekelman** [comment]
Closing: shipped in `29e4639` — feat(franken-moe): Kolen-Pollack feedback now composes with block-DFA (toy#150)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
